### PR TITLE
Issue 4099 add a preference for the preferred locale for a user.

### PR DESCRIPTION
**Pull request by zz9pzza**
This adds the preferred locale user preference, currently the ability to edit this preference is controlled by roll out.

https://code.google.com/p/otwarchive/issues/detail?id=4099

### CODE REVIEW

**Comment by sarken** [comment]
Hihi! Can you update this so it's automatically merge-able and then we'll give it one last look before merging it? Thanks!

File: app/controllers/archive_faqs_controller.rb [hunk 1]
```diff
@@ -119,7 +119,11 @@ def set_locale
     if params[:language_id] && session[:language_id] != params[:language_id]
       session[:language_id] = params[:language_id]
     end
-    @i18n_locale = session[:language_id] || I18n.default_locale
+    if current_user.present? && $rollout.active?(:set_locale_preference, current_user)
+      @i18n_locale = session[:language_id] || Locale.find(current_user.preference.prefered_locale).iso
```
Comment (sarken) [inline app/controllers/archive_faqs_controller.rb]: `prefered_locale` is spelled wrong -- it should be `preferred_locale`.

File: app/views/locales/_locale_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <h3 class="landmark heading"><%= t('.locale_heading', :default => "Locale") %></h3>
+    <dl>
+      <dt class="required"><%= f.label :name, t('.name', :default => "Name") + '*' %></dt>
+      <dd class="required"><%= f.text_field :name %></dd>
+
+      <dt class="required"><%= f.label :iso, t('.iso', :default => "ISO code") + '*' %></dt>
+      <dd class="required"><%= f.text_field :iso %></dd>
+
+      <dt class="required"><%= f.label :language_id, t('.language', :default => "Language") + '*' %></dt>
+      <dd class="required"><%= f.select(:language_id, @languages.collect {|l| [ l.name, l.id ] }) %></dd>
```
Comment (sarken) [inline app/views/locales/_locale_form.html.erb]: Can you add a space between the bracket and pipe in `{|l|`?

File: app/views/locales/_locale_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+      <dt><%= f.check_box :email_enabled %></dt>
+      <dd><%= f.label :email_enabled, t('.enable_email', :default => "Use this locale to send email")%></dd>
```
Comment (sarken) [inline app/views/locales/_locale_form.html.erb]: This might be my fault -- can you add a space before the `%` in `"Use this locale to send email")%>`?

File: app/views/locales/_locale_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+      <dt><%= f.check_box :interface_enabled %></dt>
+      <dd><%= f.label :interface_enabled, t('.enable_interface', :default => "Use this locale for the interface")%></dd>
```
Comment (sarken) [inline app/views/locales/_locale_form.html.erb]: Also probably my fault: can you add a space before the `%` in `"Use this locale for the interface")%>`?

File: app/views/preferences/index.html.erb [hunk 5]
```diff
@@ -63,28 +63,31 @@
...
+    <dd>
+      <span class="actions"><%= link_to ts('Public Site Skins'), skins_path %></span>
+      <%= f.select :skin_id, @available_skins.collect{|s| [s.title, s.id]} %>
+    </dd>
+
+    <dt><%= f.label :time_zone, ts('Your time zone: ') %></dt>
+    <dd><%= f.time_zone_select :time_zone, nil, :default => Time.zone.name %></dd>
+    
+    <% if Rails.env.test? || Rails.env.development? || $rollout.active?(:set_locale_preference, @user) %>
+      <dt><%= f.label :prefered_locale, ts('Your Locale')%> <%= link_to_help 'locale-preferences' %></dt>
```
Comment (sarken) [inline app/views/preferences/index.html.erb]: Should be spelled `:preferred_locale` Also -- and this is certainly my fault -- can you make the L in "Locale" lowercase?

File: app/views/preferences/index.html.erb [hunk 6]
```diff
@@ -63,28 +63,31 @@
...
+      <dd><%= f.select :prefered_locale, @available_locales.collect{|l| [l.name, l.id]}, :default => @preference.prefered_locale  %></dd>
```
Comment (sarken) [inline app/views/preferences/index.html.erb]: Should be `:preferred_locale` in both places on this line. Also, I think there's two spaces instead of just one before the `%>`

File: db/migrate/20141004123421_add_locale_to_preferences.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddLocaleToPreferences < ActiveRecord::Migration
+  def change
+    add_column :preferences, :prefered_locale, :integer, default: 1, null: false
```
Comment (sarken) [inline db/migrate/20141004123421_add_locale_to_preferences.rb]: Should be `:preferred_locale`

File: features/other_a/admin_tasks.feature [hunk 8]
```diff
@@ -342,24 +342,126 @@ Feature: Admin tasks
   When I edit known issues
   Then I should see "Known issue was successfully updated"
 
-  Scenario: Add a locale
+<<<<<<< HEAD
```
Comment (sarken) [inline features/other_a/admin_tasks.feature]: All of this account creation test stuff was moved to different files, so lines 345-443 should be removed.

File: public/help/locale-preferences.html [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<h4>Locale Preferences</h4>
+
+<dl id="help">
+  <dt>Set prefered locale</dt>
```
Comment (sarken) [inline public/help/locale-preferences.html]: Should be "preferred"

File: public/help/locale-preferences.html [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+  <dd>
+This preference allows you to select your prefered language for email messages that the Archive sends to you. The templates for these messages are currently being updated and translated by our volunteers. This is a work in progress; not all messages will be available in languages other than English at this time. If the template for that email has not yet been translated to your language, it will be sent in English.
```
Comment (sarken) [inline public/help/locale-preferences.html]: Should be "preferred." Also, a couple little phrasing tweaks: "This option allows you to select your preferred language for email messages the Archive sends you." and "If the template for an email has not yet been translated"

**Comment by sarken** [comment]
This looks pretty good! Just a few little nits to pick and a problem with the test to take care of and I think it will be good to go. \o/

File: app/views/locales/_locale_form.html.erb (hunk 2 above)
Comment (sarken) [inline app/views/locales/_locale_form.html.erb]: And, actually, for readability, can you use `language` rather than `l` as the variable? `l` by itself is tricky because in certain fonts (e.g. the one for writing comments here), it could be a 1, a |, or an I.

File: app/views/preferences/index.html.erb (hunk 6 above)
Comment (sarken) [inline app/views/preferences/index.html.erb]: And for readability, can you use `locale` rather than `l` as the variable? `l` by itself is tricky because in certain fonts (e.g. the one for writing comments here), it could be a 1, a |, or an I.

File: public/help/locale-preferences.html [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<h4>Locale Preferences</h4>
+
+<dl id="help">
+  <dt>Set preferred locale</dt>
+  <dd>
+This preference allows you to select your prefered language for email messages that the Archive sends to you. The templates for these messages are currently being updated and translated by our volunteers. This is a work in progress; not all messages will be available in languages other than English at this time. If the template for that email has not yet been translated to your language, it will be sent in English.
```
Comment (sarken) [inline public/help/locale-preferences.html]: One last "prefered" snuck in here.

**Comment by sarken** [comment]
I think the test failures we're seeing might actually be related to this: `Mysql2::Error: Unknown column 'locales.email_enabled' in 'where clause': SELECT `locales`.* FROM `locales`  WHERE `locales`.`email_enabled` = 1 (ActionView::Template::Error)`

File: app/views/locales/_locale_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<%= form_for(@locale, html: { class: 'post' }) do |f| %>
+  <p class="required notice">* <%= t('.required_notice', :default => "Required information") %></p>
```
Comment (ariana-paris) [inline app/views/locales/_locale_form.html.erb]: Here, you're using t(), but in the app/views/preferences/index.html.erb file below, the strings are wrapped in ts(). We probably need to decide which we want to use going forward.

File: app/views/locales/index.html.erb [hunk 13]
```diff
@@ -1,40 +1,50 @@
...
+      <% for locale in @locales %>
+        <tr>
+          <th scope="row"><%= locale.name %></th>
+          <td><%= locale.iso %></td>
+          <td><%= locale.main %></td>
+          <td><%= locale.email_enabled %></td>
+          <td><%= locale.interface_enabled %></td>
+          <td><%= locale.updated_at %></td>
+          <td class="actions">
+            <%= link_to ts('Edit'), {controller: :locales, action: :edit, id: locale.iso} %>
```
Comment (ariana-paris) [inline app/views/locales/index.html.erb]: This is a remaining ts() in a file that seems to otherwise use t()

File: db/migrate/20100929044155_add_persistence_token_to_users_and_admins.rb [hunk 14]
```diff
@@ -1,13 +0,0 @@
-class AddPersistenceTokenToUsersAndAdmins < ActiveRecord::Migration
```
Comment (ariana-paris) [inline db/migrate/20100929044155_add_persistence_token_to_users_and_admins.rb]: You seem to have scooped up a lot of database cleanup which isn't related to the locale preference and should be a separate pull request. It's not the end of the world if all these migration files need to be deleted anyway, but it's a good practice not to mix concerns when submitting a pull request by ensuring only changes that are relevant to the change at hand are committed on that branch.

Comment (zz9pzza) [inline db/migrate/20100929044155_add_persistence_token_to_users_and_admins.rb]: This is because when schema.rb was updated I should have cleaned out the migrations and I did not. Because of that I could never get the tests to work. I could have done that in another pull request but given this is due to be merged and that the clean up has to be done before any thing with migrations passes the test this was the easiest way of getting the tests to pass. Other wise I would have this pull request which is needed for two others to consider working. Which needed another pull request before its tests could work.

I tried to keep it simple :(

Comment (ariana-paris) [inline db/migrate/20100929044155_add_persistence_token_to_users_and_admins.rb]: Mixing in apples and oranges is generally not the way to simplify things, at least for reviewers, so don't do this again! *waggles finger* :) 
On this occasion, it wasn't difficult to see what was going on, so I didn't remove the "Ready to merge" label and just wanted to make sure this wasn't a mistake.
